### PR TITLE
feat: support pot format for the translation files

### LINKED ISSUE #45: po file support
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

Is there an option to extract the keys into po files instead of a json file?

### PR DESCRIPTION
**Pull request by orestisioakeimidis**
Add a new configuration option to support multiple formats (json, pot) for the translation files. The new option defaults to json and it shouldn't introduce any breaking changes.

Closes #45

## What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by orestisioakeimidis** [comment]
@shaharkazaz PR has been updated. Looking forward to your response!

**Comment by orestisioakeimidis** [comment]
@shaharkazaz Do we have any updates on this one? Thanks in advance!

**Comment by shaharkazaz** [comment]
@orestisioakeimidis hi! I'm swamped, I didn't have the time to re-review.
I'll try to find some time this weekend 🙏

File: __tests__/buildTranslationFiles.spec.ts [hunk 1]
```diff
@@ -54,36 +56,55 @@ type TranslationCategory =
...
+    outputFormat
+  );
 }
 
 function removeI18nFolder(type: TranslationCategory) {
   fs.removeSync(`./${sourceRoot}/${type}/i18n`);
 }
 
-describe('buildTranslationFiles', () => {
+const formats: ['json' | 'pot'][] = [['json'], ['pot']];
```
Comment (shaharkazaz) [inline __tests__/buildTranslationFiles.spec.ts:96]: Why is this an array of arrays? 🤔

File: __tests__/buildTranslationFiles.spec.ts [hunk 2]
```diff
@@ -338,17 +373,16 @@ describe('buildTranslationFiles', () => {
 
       beforeEach(() => removeI18nFolder(type));
 
-      it('show work with unflat true', () => {
+      it('should work with unflat true', () => {
```
Comment (shaharkazaz) [inline __tests__/buildTranslationFiles.spec.ts:376]: 🔥

File: __tests__/buildTranslationFiles.spec.ts [hunk 3]
```diff
@@ -109,8 +130,8 @@ describe('buildTranslationFiles', () => {
           expected[nonNumericKey] = defaultValue;
         });
 
-        buildTranslationFiles(config);
-        assertTranslation({ type, expected });
+        buildTranslationFiles({ ...config, outputFormat });
```
Comment (shaharkazaz) [inline __tests__/buildTranslationFiles.spec.ts:133]: This seems very repetitive, maybe we should wrap it with a function that expects the config?
Something like: 
```ts
createTranslations(config)

function createTranslations(config) {
 buildTranslationFiles({ ...config, outputFormat });
}
```
What do you think?

File: src/config.ts [hunk 4]
```diff
@@ -55,5 +55,8 @@ export function defaultConfig(
 
     // The path for the root translation files (for example: assets/i18n)
     translationsPath: 'assets/i18n',
+
+    // The output format (`json`, `pot`) of the translation files (defaults to `json`)
```
Comment (shaharkazaz) [inline src/config.ts:59]:
```suggestion
    // The output format of the translation files (`json`, `pot`)
```

File: src/keys-builder/build-translation-file.ts [hunk 5]
```diff
@@ -12,27 +11,15 @@ export interface FileAction {
 export function buildTranslationFile(
   path: string,
   translation = {},
-  replace = false
+  replace = false,
+  outputFormat
```
Comment (shaharkazaz) [inline src/keys-builder/build-translation-file.ts:15]: I think we need to convert the params to an object

File: src/keys-builder/utils/create-translation.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import { unflatten } from 'flat';
+import { po } from 'gettext-parser';
+
+import { getConfig } from '../../config';
+import { mergeDeep, stringify } from '../../utils/object.utils';
+
+function handleReplace(currentTranslation, translation, replace) {
```
Comment (shaharkazaz) [inline src/keys-builder/utils/create-translation.ts:7]:
```suggestion
function resolveTranslation({currentTranslation, translation, replace}) {
```
We might have more logic here in the future.

File: src/keys-builder/utils/create-translation.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+function createJson(currentTranslation, translation, replace) {
+  if (getConfig().unflat) {
+    translation = unflatten(translation, { object: true });
+  }
+
+  const value = handleReplace(currentTranslation, translation, replace);
+
+  return stringify(value);
```
Comment (shaharkazaz) [inline src/keys-builder/utils/create-translation.ts:20]:
```suggestion
  const resolved = resolveTranslation({currentTranslation, translation, replace});

  return stringify(resolved);
```

File: src/keys-builder/utils/create-translation.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        'content-type': 'text/plain; charset=utf-8',
+        'content-transfer-encoding': '8bit',
+      },
+      translations: {
+        '': Object.entries(value).reduce(
+          (acc, [msgid, msgstr]) => ({
+            ...acc,
+            [msgid]: { msgid, msgstr },
+          }),
+          {} as any
```
Comment (shaharkazaz) [inline src/keys-builder/utils/create-translation.ts:40]: why `as any`?

File: src/keys-builder/utils/create-translation.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+export function createTranslation(
+  currentTranslation,
+  translation,
+  replace,
+  outputFormat
+) {
+  const compilers = {
+    json: createJson,
+    pot: createPot,
+  };
```
Comment (shaharkazaz) [inline src/keys-builder/utils/create-translation.ts:56]: Please move outside the function, no need to recreate each time 😄

File: src/keys-builder/utils/create-translation.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      },
+    })
+    .toString('utf8');
+}
+
+export function createTranslation(
+  currentTranslation,
+  translation,
+  replace,
+  outputFormat
```
Comment (shaharkazaz) [inline src/keys-builder/utils/create-translation.ts:51]: This needs to be an object, same for the `createJson` & `createPot`

File: src/keys-builder/utils/get-current-translation.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  } catch {
+    return {};
+  }
+}
+
+export function getCurrentTranslation(path, outputFormat) {
+  const parsers = {
+    json: getJson,
+    pot: getPot,
+  };
```
Comment (shaharkazaz) [inline src/keys-builder/utils/get-current-translation.ts:40]: Move outside the function

File: src/keys-builder/utils/get-current-translation.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        (acc, key) => ({
+          ...acc,
+          [key]: parsed.translations[''][key].msgstr.pop(),
+        }),
+        {}
+      );
+
+    return getConfig().unflat ? unflatten(value, { object: true }) : value;
+  } catch {
+    return {};
```
Comment (shaharkazaz) [inline src/keys-builder/utils/get-current-translation.ts:32]: I think we might want to send a warning here? what's the case you are protecting?

File: src/keys-builder/utils/get-current-translation.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+import { po } from 'gettext-parser';
+import { unflatten } from 'flat';
+
+import { getConfig } from '../../config';
+
+function getJson(path) {
+  return fsExtra.readJsonSync(path, { throws: false }) || {};
+}
+
+function getPot(path) {
```
Comment (shaharkazaz) [inline src/keys-builder/utils/get-current-translation.ts:11]: They are under an object called `parser`, they should be called: `parseJson` & `parsePot`

File: src/keys-builder/utils/scope.utils.ts [hunk 14]
```diff
@@ -15,3 +15,8 @@ export function getScopes() {
 export function hasScope(scope: string) {
   return scopeToAlias.hasOwnProperty(scope);
 }
+
+export function resetScopes() {
+  scopeToAlias = {};
+  aliasToScope = {};
+}
```
Comment (shaharkazaz) [inline src/keys-builder/utils/scope.utils.ts:22]: Correct me if I'm wrong, this is something related just to tests right?
Also, what was the case that made you add this?

File: src/keys-detective/compare-keys-to-files.ts [hunk 15]
```diff
@@ -65,14 +71,20 @@ export function compareKeysToFiles({
...
     }
   }
 
   for (const { files, keys, scope, translationPath } of result) {
     for (const filePath of files) {
-      const { lang } = getScopeAndLangFromPath(filePath, translationPath);
+      const { lang } = getScopeAndLangFromPath(
+        filePath,
+        translationPath,
+        outputFormat
```
Comment (shaharkazaz) [inline src/keys-detective/compare-keys-to-files.ts:86]: convert to an object

**Review by shaharkazaz — COMMENTED** [review]
Greate work! 🔥

**Comment by shaharkazaz** [comment]
@orestisioakeimidis Just making sure you saw the comments 😄

**Comment by orestisioakeimidis** [comment]
Hey @shaharkazaz, yes I saw them. Thanks for picking it up! I'll try to have a look this week.

File: __tests__/buildTranslationFiles.spec.ts (hunk 1 above)
Comment (orestisioakeimidis) [reply __tests__/buildTranslationFiles.spec.ts:96]: Leftover, I will change it to a simple array.

**Comment by orestisioakeimidis** [comment]
Pushed new changes and responded to a few comments @shaharkazaz.

File: src/keys-builder/utils/scope.utils.ts (hunk 14 above)
Comment (shaharkazaz) [reply src/keys-builder/utils/scope.utils.ts:22]: If it's only test-related please resolve this from within the test's scope, this shouldn't be part of the source code.

**Comment by shaharkazaz** [comment]
@orestisioakeimidis not sure if you noticed, I left new comments 🙂

**Comment by orestisioakeimidis** [comment]
@shaharkazaz I've seen them, but didn't have the time to work on it yet. Will try to do it asap. Thanks for checking!

**Comment by shaharkazaz** [comment]
@orestisioakeimidis Any chance you can pick this up? it's really almost done! I want to merge this 🔥

**Comment by orestisioakeimidis** [comment]
I'll do it tomorrow @shaharkazaz. I've been feeling sick last week and didn't have the chance.

**Comment by shaharkazaz** [comment]
@orestisioakeimidis sure thing man! Hope everything is well 🙏

File: src/keys-builder/utils/scope.utils.ts (hunk 14 above)
Comment (orestisioakeimidis) [reply src/keys-builder/utils/scope.utils.ts:22]: Honestly, I can't think of a way to resolve this within the test's scope. We need to reset the scopes and the only way I can think of is to expose this method for doing so.

We could mark this method as a method only meant for internal use or something similar.

Unless you have some other idea maybe...?

**Comment by orestisioakeimidis** [comment]
@shaharkazaz pushed a change and responded in one of the comments.

File: src/keys-builder/utils/create-translation.ts (hunk 8 above)
Comment (shaharkazaz) [reply src/keys-builder/utils/create-translation.ts:48]: I think you forgot to remove it 🙂

File: src/keys-builder/utils/create-translation.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  json: createJson,
+  pot: createPot,
+};
+
+export function createTranslation({
+  currentTranslation,
+  translation,
+  replace,
+  outputFormat,
+}) {
```
Comment (shaharkazaz) [inline src/keys-builder/utils/create-translation.ts:65]: Missing types in all these functions

File: src/keys-builder/utils/get-current-translation.ts [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        {}
+      );
+
+    return getConfig().unflat ? unflatten(value, { object: true }) : value;
+  } catch (e) {
+    if (e.code === 'ENOENT') {
+      return {};
+    }
+
+    console.warn('Something is wrong with the provided file', e.message);
```
Comment (shaharkazaz) [inline src/keys-builder/utils/get-current-translation.ts:36]: I think the message should include the `path`

File: src/keys-builder/utils/get-current-translation.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+  }
+}
+
+const parsers = {
+  json: parseJson,
+  pot: parsePot,
+};
+
+export function getCurrentTranslation({ path, outputFormat }) {
```
Comment (shaharkazaz) [inline src/keys-builder/utils/get-current-translation.ts:45]: Missing types in all the functions here

File: src/keys-builder/utils/scope.utils.ts (hunk 14 above)
Comment (shaharkazaz) [reply src/keys-builder/utils/scope.utils.ts:22]: It's ok for since we aren't going to expose this 👍

**Comment by shaharkazaz** [comment]
@orestisioakeimidis Left a few last comments, note that you need to update from master 🙂

**Comment by orestisioakeimidis** [comment]
@shaharkazaz pushed changes and updated from master!

**Comment by shaharkazaz** [comment]
@orestisioakeimidis Great work 👍

**Comment by shaharkazaz** [comment]
@orestisioakeimidis Note that I made some refactors in https://github.com/ngneat/transloco-keys-manager/commit/d2c84d974a55c4fea3deb86a33bcd3feb259f272 and renamed the input to **file format**.
I also made several type changes to make the code stricter. You are welcome to take a look 🙂

**Comment by orestisioakeimidis** [comment]
It looks great @shaharkazaz! Only one thing you missed is the `outputFormat` name in the README file. Make sure to fix it! :)

**Pull request merged**